kafka/metadata.go

```diff
@@ -156,3 +156,29 @@ func queryWatermarkOffsets(H Handle, topic string, partition int32, timeoutMs in
 	high = int64(cHigh)
 	return low, high, nil
 }
+
+// getWatermarkOffsets returns the clients cached low and high offsets for the given topic
+// and partition.
+func getWatermarkOffsets(H Handle, topic string, partition int32) (low, high int64, err error) {
+	h := H.gethandle()
+
+	ctopic := C.CString(topic)
+	defer C.free(unsafe.Pointer(ctopic))
+
+	var cLow, cHigh C.int64_t
+
+	e := C.rd_kafka_get_watermark_offsets(h.rk, ctopic, C.int32_t(partition),
+		&cLow, &cHigh)
+	if e != C.RD_KAFKA_RESP_ERR_NO_ERROR {
+		return 0, 0, newError(e)
+	}
+
+	if cLow != C.RD_KAFKA_OFFSET_INVALID {
```

Comment:
I don't think these two if statements should be there, since it will convert an Invalid offset into a proper one (0).
Just do the cast.

Reply:
If I remove them, my current test will fail with

```
integration_test.go:487: QueryWatermarkOffsets low[0] does not equal GetWatermarkOffsets low[-1001]
```

`QueryWatermarkOffsets` correctly returns 0 for the low watermark but `GetWatermarkOffsets` will return `RD_KAFKA_OFFSET_INVALID` (1001).

Is that the behavior that we want ?

Reply:
Yes, that's the correct behaviour.
It must not default to offset 0, which is a valid offset, but not necessarily the low offset of the partition (after log segment deletion).

Reply:
To fix your test:
Add statistics.interval.ms=100 and a short sleep (200ms) after the first message has been receieved to make sure the low offset is populated.

Reply:
I tried that by doing something like

```
	seenMessage, seenStats := false, false
	// Wait for messages to be received so that we know the watermark offsets have been delivered
	// with the fetch response
	for ev := range c.Events() {
		if _, ok := ev.(*Message); ok {
			seenMessage = true
		}
		if _, ok := ev.(*Stats); ok {
			seenStats = true
		}
		if seenMessage && seenStats {
			break
		}
	}
```

But the low watermark is still invalid.  Even sleeping for 5s doesn't help.

Per my comment above and based on a quick glance through the code, I couldn't find anything that actually sets the specific `rktp_lo_offset` field

Reply:
It is set here, in the Offset consumer_lag response callback, which is triggered every statistics.interval.ms.
https://github.com/edenhill/librdkafka/blob/d306f5f7ef2aea04cdb9e7146796efd4a28d49f1/src/rdkafka_partition.c#L106

But it only fires for assigned partitions, so the sleep needs to be done after the group assignment has kicked in, and the safest way is to sleep after seeing the first message.

Reply:
OK I believe that figured out why it wasn't working.  It looks like minimum interval to begin querying the lower watermark offset is 10s per [this block](https://github.com/edenhill/librdkafka/blob/d306f5f7ef2aea04cdb9e7146796efd4a28d49f1/src/rdkafka_partition.c#L225-L236).

If I sleep for at least 10s, the code works as expected but this isn't ideal.  I'm inclined to avoid checking the low offset in this test.  Thoughts ?

Reply:
You're right, 10s! :)

I'm fine with skipping the low offset check, yeah.